### PR TITLE
Add a rotating buffer functionality to the s3 driver

### PR DESCRIPTION
**Pull request by AndreyKostov**

### CODE REVIEW

File: storagedriver/s3/s3.go [hunk 1]
```diff
@@ -284,6 +284,8 @@ func (d *Driver) WriteStream(path string, offset int64, reader io.Reader) (total
 
 	partNumber := 1
 	bytesRead := 0
+	putErrChan := make(chan error, 1)
```
Comment (stevvooe) [inline storagedriver/s3/s3.go]: Try this: change this to:

```
var putErrChan chan error
```

File: storagedriver/s3/s3.go [hunk 2]
```diff
@@ -356,16 +363,28 @@ func (d *Driver) WriteStream(path string, offset int64, reader io.Reader) (total
...
 
-		if bytesRead > 0 {
-			part, err = multi.PutPart(int(partNumber), bytes.NewReader(buf[0:int64(bytesRead)+from]))
-			if err != nil {
-				return err
-			}
-
-			parts = append(parts, part)
-			partNumber++
+		putErr := <-putErrChan
```
Comment (stevvooe) [inline storagedriver/s3/s3.go]: Then here:

```
var putErr error
if putErrChan == nil {
    putErrChan = make(chan err)
    putErr = nil
} else {
    putErr = <-putErrChan
}
```

**Comment by stevvooe** [comment]
I think this is looking okay. Did you run the unit tests with the race detector enabled?

**Comment by AndreyKostov** [comment]
There was a small bug with overwriting err (I had missed it). We should be good now (ran full test suite).

**Comment by dmp42** [comment]
LGTM

**Comment by stevvooe** [comment]
LGTM!

**Pull request merged**